### PR TITLE
NIAD-1593: Immunization: The userSelected code is not being selected for the site code

### PR DESCRIPTION
**Pull request by bbell-kainos**

### CODE REVIEW

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java [hunk 1]
```diff
@@ -177,12 +178,25 @@ private String buildExpirationDatePertinentInformation(Immunization immunization
...
+        String siteUserSelected = immunization.getSite().getCoding().stream()
+            .filter(Coding::hasUserSelected)
+            .findFirst()
+            .map(Coding.class::cast)
+            .map(Coding::getDisplay)
+            .filter(StringUtils::isNotBlank)
+            .map(value -> SITE + value)
+            .orElse(StringUtils.EMPTY);
+
+        if (StringUtils.isBlank(siteUserSelected)) {
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:190]: This whole thing can be rewritten:
```
return immunization.getSite().getCoding().stream()
    .filter(Coding::hasUserSelected)
    .findFirst()
    .map(Coding.class::cast)
    .map(Coding::getDisplay)
    .filter(StringUtils::isNotBlank)
    .map(value -> SITE + value)
    .or(() -> Optional.of(immunization)
        .filter(Immunization::hasSite)
        .map(Immunization::getSite)
        .map(CodeableConceptMappingUtils::extractTextOrCoding)
        .map(value -> SITE + value))
    .orElse(StringUtils.EMPTY);
```

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java (hunk 1 above)
Comment (bbell-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java]: Original code has been extracted into a new method in 'CodeableConceptMappingUtils'

**Review by ElliotB-96 — DISMISSED** [review]

**Review by rorysproule1 — DISMISSED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/CodeableConceptMappingUtils.java [hunk 2]
```diff
@@ -16,6 +16,20 @@
         }
     }
 
+    public static Optional<String> extractUserSelectedTextOrCoding(CodeableConcept codeableConcept) {
+        var userSelected = codeableConcept.getCoding().stream()
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/CodeableConceptMappingUtils.java:20]: Can be:
```
return codeableConcept.getCoding().stream()
    .filter(Coding::hasUserSelected)
    .findFirst()
    .map(Coding.class::cast)
    .map(Coding::getDisplay)
    .or(() -> CodeableConceptMappingUtils.extractTextOrCoding(codeableConcept));

```

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java [hunk 3]
```diff
@@ -177,12 +177,12 @@ private String buildExpirationDatePertinentInformation(Immunization immunization
...
 
     private String buildSitePertinentInformation(Immunization immunization) {
-        Optional<String> site = Optional.empty();
-        if (immunization.hasSite()) {
-            CodeableConcept siteObject = immunization.getSite();
-            site = CodeableConceptMappingUtils.extractTextOrCoding(siteObject);
-        }
-        return site.map(value -> SITE + value).orElse(StringUtils.EMPTY);
+        StringBuilder siteInformation = new StringBuilder(StringUtils.EMPTY);
+        CodeableConceptMappingUtils.extractUserSelectedTextOrCoding(immunization.getSite()).ifPresent(siteValue -> {
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:181]: Don't you need to check for `hasSite()` ?

Comment (bartek-sarul) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:181]: It's more readable when `ifPresent` is on new line

Comment (bartek-sarul) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java:181]: Anyway why not do this:
```
return Optional.of(immunization)
    .filter(Immunization::hasSite)
    .map(Immunization::getSite)
    .map(siteValue -> SITE + siteValue)
    .orElse(StringUtils.EMPTY);
```

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/CodeableConceptMappingUtils.java (hunk 2 above)
Comment (bbell-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/utils/CodeableConceptMappingUtils.java]: Addressed

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java (hunk 3 above)
Comment (bbell-kainos) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ImmunizationObservationStatementMapper.java]: Addressed

**Review by bartek-sarul — APPROVED** [review]

**Review by rorysproule1 — APPROVED** [review]

**Pull request merged**